lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,406 @@
+	cfg.SetDefaults()
+	return &Buffer[T]{
+		cfg:    cfg,
+		notify: make(chan struct{}),
+		ring:   make([]entry[T], cfg.Capacity),
+	}
+}
+
+// Close permanently closes the fanout buffer. Note that all cursors are termianted immediately
+// and may therefore miss items that had not been read at the time Close is called.
+func (b *Buffer[T]) Close() {
+	var notify chan struct{}
+	b.write(func() {
+		notify = b.notify
+		b.notify = make(chan struct{})
+		b.closed = true
+	})
+	close(notify)
+}
+
+// NewCursor gets a new cursor into the buffer. Cursors *must* be closed as soon as they are
+// no longer being read from. Stale cursors may cause performance degredation.
+func (b *Buffer[T]) NewCursor() *Cursor[T] {
+	var c uint64
+	b.write(func() {
+		if b.closed {
+			return
+		}
+		c = b.pos + b.len
+		b.cursors++
+	})
+
+	return &Cursor[T]{
+		buf:    b,
+		cursor: c,
+	}
+}
+
+// Append appends to the buffer and wakes all dormant cursors.
+func (b *Buffer[T]) Append(items ...T) {
+
+	var notify chan struct{}
+
+	b.write(func() {
+		if b.closed {
+			panic("Append called on closed fanout buffer")
+		}
+		for _, item := range items {
+			// ensure there is at least one free slot in ring
+			b.ensureSlot()
+
+			// insert new entry into ring
+			idx := int((b.pos + b.len) % b.cfg.Capacity)
+			b.ring[idx].item = item
+			b.ring[idx].wait.Store(b.cursors)
+			b.len++
+		}
+
+		notify = b.notify
+		b.notify = make(chan struct{})
+	})
+
+	// notify all waiting cursors. we do this outside of the
+	// lock to (theoretically) reduce potential contention.
+	close(notify)
+}
+
+// ensure that there is at least one open slot in the ring. expires and overflows items as needed.
+func (b *Buffer[T]) ensureSlot() {
+	// reset overflow if it is past its grace period
+	if !b.overflowStart.IsZero() && b.cfg.Clock.Now().After(b.overflowStart.Add(b.cfg.GracePeriod)) {
+		b.overflow = nil
+	}
+
+	// trim items from overflow that have been seen by all cursors
+	var clearOverflowTo int
+	for i := 0; i < len(b.overflow); i++ {
+		clearOverflowTo = i
+		if b.overflow[i].wait.Load() > 0 {
+			break
+		}
+	}
+	b.overflow = b.overflow[clearOverflowTo:]
+
+	// clear overflow start time if overflow is empty
+	if len(b.overflow) == 0 {
+		b.overflowStart = time.Time{}
+	}
+
+	// clear items from ring that have been seen by all cursors
+	for b.len > 0 && b.ring[int(b.pos%b.cfg.Capacity)].wait.Load() == 0 {
+		b.ring[int(b.pos%b.cfg.Capacity)] = entry[T]{}
+		b.pos++
+		b.len--
+	}
+
+	if b.len < b.cfg.Capacity {
+		// we have at least one free slot in ring, no need for overflow
+		return
+	}
+
+	// copy oldest entry to overflow. note the that it would actually be safe to just append the
+	// entire `entry` value rather than this append -> load -> store pattern that we follow
+	// here, but 'go vet' doesn't understand our locking semantics so we need to avoid implicitly
+	// copying the atomic, even when the write lock is held.
+	b.overflow = append(b.overflow, entry[T]{
```

Comment:
I _could_ add a config option for a limit, but honestly for the main intended usecase (backend event fanout), I don't think I'd want to enforce one.  I don't think there's any specific value that would both be large enough that it wouldn't cause issues during legitimate usage, while also being small enough to prevent OOMs